sql : alter sequence's increment and the results of 'select nextval()' were not as expected

mosquito2333 opened the pull request:
sql : alter sequence's increment and the results of 'select nextval()' were not as expected

If the sequenceVal exceeded the old MinValue or MaxValue, we must set sequenceVal to the last valid one.

1. Values is just init, when the sequence is altered before it has ever been used.
In this case, last valid sequenceVal is oldStart.
Set the initial value to the oldStartValue(just like pg), when last valid sequenceVal in [newMinValue, newMaxValue].

2. When the sequenceVal out of range result from increment, report an error.

Fixes #52552

sql : sequence value out of bounds

    Previously, an additional out-of-range sequence value may have been
    generated.
    The original criterion was whether the previous endSequence value exceeded the limit.
    However, when the current endSequence is 1, the increment is 2 and the MaxValue is 2,
    the previous endSequence value is 1 and the program does not report an error.
    However, when the current endSequenceValue 3 exceeds the limit, an error should be reported

    In incrementSequenceUsingCache, the criterion for determining whether a cache's data is eligible is modified to whether the first data that may be stored in the cache exceeds the limit.

Fixes #74127.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74321)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mosquito2333)

___
*[pkg/sql/sequence.go, line 156 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTn8TY8DmzijwVWn6t:-MsTn8TY8DmzijwVWn6u:b-r2u7df) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/sequence.go#L156)):*
> ```Go
> 		}
> 
> 		incrementAmount = seqOpts.Increment
> ```

I don't see any value in this assignment: we're not modifying `seqOpts`. Your variable name is one character fewer. If the verbosity is really getting to you, I can see renaming `seqOpts` to `opts`. 
___
*[pkg/sql/sequence.go, line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTlqE40b_kWqeGkz3t:-MsTlqE40b_kWqeGkz3u:ba87ooq) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/sequence.go#L158)):*
> ```Go
> 		incrementAmount = seqOpts.Increment
> 		lastEndValue := endValue - incrementAmount*cacheSize
> 		beginValue := endValue - incrementAmount*(cacheSize-1)
> ```

I find these last two variable names confusing. They're confusing because it's not clear what `begin` means in this context. `endValue` here is intended to mean: the value of the sequence stored in KV after the operation. However, `begin` here doesn't mean the value before the operation. How about renaming the variables to `postIncrement`, `preIncrement`, `nextVal` and then reworking the order of condition checking. Consider the below snippet: 

```go
	fetchNextValues := func() (nextVal, incrementAmount, sizeOfCache int64, err error) {
		seqValueKey := p.ExecCfg().Codec.SequenceKey(uint32(descriptor.GetID()))

		cacheIncrement := seqOpts.Increment * cacheSize
		postIncrement, err := kv.IncrementValRetryable(ctx, p.ExecCfg().DB, seqValueKey, cacheIncrement)
    		if err != nil {
			if errors.HasType(err, (*roachpb.IntegerOverflowError)(nil)) {
				return 0, 0, 0, boundsExceededError(descriptor)
			}
			return 0, 0, 0, err
		}
          
		preIncrement := endValue - seqOpts.Increment*cacheSize
		nextVal = preIncrement + seqOpts.Increment
```

Now, invert the condition to aid readability:
```go
		// The postIncrement value is in-bounds, this implies that nextVal and all interceding values are too.
		if postIncrement < seqOpts.MaxValue && postIncrement > seqOpts.MinValue {
			return nextVal, seqOpts.Increment, cacheSize, nil
		}
		// The postIncrement value is out-of-bounds, find the portion of the cached sequence which is valid, if any.
                positiveIncrement := seqOpts.Increment > 0 
                if (positiveIncrement && nextVal > seqOpts.MaxValue) || 
			(!positiveIncrement && nextVal < seqOpts.MinValue) {
			return 0, 0, 0, boundsExceededError(descriptor)
                }
		// Otherwise, some values between the limit and the value prior to incrementing can be cached.
		// Figure out how many.
                var limit int64
                switch positiveIncrement {
                case true:
                	limit = seqOpts.MaxValue
                case false:
                	limit = seqOpts.MinValue
                }
		sizeOfCache = abs(limit-lastEndValue) / abs(seqOpts.Increment)
                return nextVal, seqOpts.Increment, sizeOfCache, nil
}
```

Move `abs` out of here
```go
		abs := func(i int64) int64 {
			if i < 0 {
				return -i
			}
			return i
		}
```

Also, move `nextVal`'s definition down into the conditional. 



_Code quote:_
```Go
		lastEndValue := endValue - incrementAmount*cacheSize
		beginValue := endValue - incrementAmount*(cacheSize-1)
```
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTl7FyDLAlf72DDPa6:-MsTl7FyDLAlf72DDPa7:b-8gjtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/logictest/testdata/logic_test/sequences#L2034)):*
> ```
> 
> ###### check the value out of bounds
> ## cache = 1
> ```

nit: please formulate these comments as sentences. 



_Code quote:_
```
###### check the value out of bounds
## cache = 1
```
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2048 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTl3E_-KVV-UfqQNJk:-MsTl3E_-KVV-UfqQNJl:b-mgqs2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/logictest/testdata/logic_test/sequences#L2048)):*
> ```
> 9
> 
> query error pq: nextval\(\): reached maximum value of sequence "customer_seq_check_cache_and_bounds_1" \(10\)
> ```

can you assert the `pgcode` with the `pgcode` directive for the error cases?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2056 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTlURj1YPSVZVdJOiq:-MsTlURj1YPSVZVdJOir:box43tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/logictest/testdata/logic_test/sequences#L2056)):*
> ```
> 12
> 
> ## cache = 2
> ```

why is the comment here `cache = 2` and then the statement says `CACHE 5`. 


<!-- Sent from Reviewable.io -->

mosquito2333 commented:
A question arises, how can I get the last valid sequence value?

mosquito2333 commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mosquito2333)
> 
> _[pkg/sql/sequence.go, line 156 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTn8TY8DmzijwVWn6t:-MsTn8TY8DmzijwVWn6u:b-r2u7df) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/sequence.go#L156)):_
> 
> > ```go
> > 		}
> > 
> > 		incrementAmount = seqOpts.Increment
> > ```
> 
> I don't see any value in this assignment: we're not modifying `seqOpts`. Your variable name is one character fewer. If the verbosity is really getting to you, I can see renaming `seqOpts` to `opts`.
> 
> _[pkg/sql/sequence.go, line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MsTlqE40b_kWqeGkz3t:-MsTlqE40b_kWqeGkz3u:ba87ooq) ([raw file](https://github.com/cockroachdb/cockroach/blob/09fbbab9c6e59637d1be5639e6c30d306dffc124/pkg/sql/sequence.go#L158)):_
> 
> > ```go
> > 		incrementAmount = seqOpts.Increment
> > 		lastEndValue := endValue - incrementAmount*cacheSize
> > 		beginValue := endValue - incrementAmount*(cacheSize-1)
> > ```
> 
> I find these last two variable names confusing. They're confusing because it's not clear what `begin` means in this context. `endValue` here is intended to mean: the value of the sequence stored in KV after the operation. However, `begin` here doesn't mean the value before the operation. How about renaming the variables to `postIncrement`, `preIncrement`, `nextVal` and then reworking the order of condition checking. Consider the below snippet:
> 
> ```go
> 	fetchNextValues := func() (nextVal, incrementAmount, sizeOfCache int64, err error) {
> 		seqValueKey := p.ExecCfg().Codec.SequenceKey(uint32(descriptor.GetID()))
> 
> 		cacheIncrement := seqOpts.Increment * cacheSize
> 		postIncrement, err := kv.IncrementValRetryable(ctx, p.ExecCfg().DB, seqValueKey, cacheIncrement)
>     		if err != nil {
> 			if errors.HasType(err, (*roachpb.IntegerOverflowError)(nil)) {
> 				return 0, 0, 0, boundsExceededError(descriptor)
> 			}
> 			return 0, 0, 0, err
> 		}
>           
> 		preIncrement := endValue - seqOpts.Increment*cacheSize
> 		nextVal = preIncrement + seqOpts.Increment
> ```
> 
> Now, invert the condition to aid readability:
> 
> ```go
> 		// The postIncrement value is in-bounds, this implies that nextVal and all interceding values are too.
> 		if postIncrement < seqOpts.MaxValue && postIncrement > seqOpts.MinValue {
> 			return nextVal, seqOpts.Increment, cacheSize, nil
> 		}
> 		// The postIncrement value is out-of-bounds, find the portion of the cached sequence which is valid, if any.
>                 positiveIncrement := seqOpts.Increment > 0 
>                 if (positiveIncrement && nextVal > seqOpts.MaxValue) || 
> 			(!positiveIncrement && nextVal < seqOpts.MinValue) {
> 			return 0, 0, 0, boundsExceededError(descriptor)
>                 }
> 		// Otherwise, some values between the limit and the value prior to incrementing can be cached.
> 		// Figure out how many.
>                 var limit int64
>                 switch positiveIncrement {
>                 case true:
>                 	limit = seqOpts.MaxValue
>                 case false:
>                 	limit = seqOpts.MinValue
>                 }
> 		sizeOfCache = abs(limit-lastEndValue) / abs(seqOpts.Increment)
>                 return nextVal, seqOpts.Increment, sizeOfCache, nil
> }
> ```
We can't report an error directly, otherwise the eligible values in this cache will also report an error. 
Such as:
the first cache will contain 1,-1,-3,-5,-7
1,-1,-3 are in[-2,2],but -5,-7 are not.
```go
 if (positiveIncrement && nextVal > seqOpts.MaxValue) || 
> 			(!positiveIncrement && nextVal < seqOpts.MinValue) {
> 			return 0, 0, 0, boundsExceededError(descriptor)
>                 }
```
If use this,an error will report when the result of nextval is 1
```sql
CREATE SEQUENCE customer_seq_check_cache_and_bounds_2 MINVALUE -2 MAXVALUE 2 START WITH 1 CACHE 5 INCREMENT BY -2

query I
SELECT nextval('customer_seq_check_cache_and_bounds_2')
----
1

query I
SELECT nextval('customer_seq_check_cache_and_bounds_2')
----
-1

query error pq: nextval\(\): reached minimum value of sequence "customer_seq_check_cache_and_bounds_2" \(-2\)
SELECT nextval('customer_seq_check_cache_and_bounds_2')

```

mosquito2333 commented:
How do I use pgcode for the error messages in testing files? I can't find an example, can you give me some help?

mosquito2333 commented:
- The parameter name has been changed in alter_sequence.go.
- How do I use pgcode for the error messages in testing files? I can't find an example, can you give me some help?
- No changes to the procedure in sequence.go

mosquito2333 commented:
Now in assignSequenceOptions(), when checking the descriptor of sequence, the old start parameter is compared with the new bounds.
 In fact, we should compare the last valid sequence value with the new bounds after assignSequenceOptions() when alter sequence.
 If a new start parameter is passed, start is used, otherwise the last valid sequence value is used.

I found a pr, but I haven't understood it.
@rafiss 
https://github.com/cockroachdb/cockroach/pull/57339/files

rafiss commented:
@mosquito2333 
> How do I use pgcode for the error messages in testing files? I can't find an example, can you give me some help?

Here is an example: https://github.com/cockroachdb/cockroach/blob/fa91c719392104a9d288db44b6afad6deab1ef87/pkg/sql/logictest/testdata/logic_test/alter_table#L46

mosquito2333 commented:
> @mosquito2333
> 
> > How do I use pgcode for the error messages in testing files? I can't find an example, can you give me some help?
> 
> Here is an example:
> 
> https://github.com/cockroachdb/cockroach/blob/fa91c719392104a9d288db44b6afad6deab1ef87/pkg/sql/logictest/testdata/logic_test/alter_table#L46

Thank you! Revisions completed.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mosquito2333)

___
*[-- commits, line 28 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl3l1lDCfp4tB1xGcV):*
the commit message should have a release note:
```
Release note (bug fix): A bug with ALTER SEQUENCE that allowed a sequence to return values
that are out-of-range has been fixed.
```
___
*[pkg/sql/alter_sequence.go, line 157 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl5sOXBNLdm0OWloXD:-Msl5sOY2q76BGshcIUe:b-fotbq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/96acd7d758f351bc69dd307a52a4684bde10933d/pkg/sql/alter_sequence.go#L157)):*
> ```Go
> 	}
> 	// Clear out the cache and update the last value.
> 	params.p.sessionDataMutatorIterator.applyOnEachMutator(func(m sessionDataMutator) {
> ```

should the "clear cache and record latest val" be done after writeSchemaChange?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2052 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl39m_-gUMPhnvjQtE:-Msl39m_-gUMPhnvjQtF:bxn49z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/db37312c44ebebe741629b7480a2f2bbaaf004a9/pkg/sql/logictest/testdata/logic_test/sequences#L2052)):*
> ```
> 
> query I
> SELECT last_value from customer_seq_check_cache_and_bounds_1
> ```

i noticed that in postgres, this query would return `9`. can we match that behavior?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2076 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl3P8_EMqcAIIOr5Fa:-Msl3P8a2P6XWwaX7xmv:b-i0ha0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/db37312c44ebebe741629b7480a2f2bbaaf004a9/pkg/sql/logictest/testdata/logic_test/sequences#L2076)):*
> ```
> SELECT last_value from customer_seq_check_cache_and_bounds_2
> ----
> -17
> ```

`-17` is a pretty surprising result to get here. in postgres, this query returns `-1`


<!-- Sent from Reviewable.io -->

mosquito2333 commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mosquito2333)
> 
> _[-- commits, line 28 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl3l1lDCfp4tB1xGcV):_ the commit message should have a release note:
> 
> ```
> Release note (bug fix): A bug with ALTER SEQUENCE that allowed a sequence to return values
> that are out-of-range has been fixed.
> ```
> 
> _[pkg/sql/alter_sequence.go, line 157 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl5sOXBNLdm0OWloXD:-Msl5sOY2q76BGshcIUe:b-fotbq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/96acd7d758f351bc69dd307a52a4684bde10933d/pkg/sql/alter_sequence.go#L157)):_
> 
> > ```go
> > 	}
> > 	// Clear out the cache and update the last value.
> > 	params.p.sessionDataMutatorIterator.applyOnEachMutator(func(m sessionDataMutator) {
> > ```
> 
> should the "clear cache and record latest val" be done after writeSchemaChange?
> 
> _[pkg/sql/logictest/testdata/logic_test/sequences, line 2052 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl39m_-gUMPhnvjQtE:-Msl39m_-gUMPhnvjQtF:bxn49z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/db37312c44ebebe741629b7480a2f2bbaaf004a9/pkg/sql/logictest/testdata/logic_test/sequences#L2052)):_
> 
> > ```
> > 
> > query I
> > SELECT last_value from customer_seq_check_cache_and_bounds_1
> > ```
> 
> i noticed that in postgres, this query would return `9`. can we match that behavior?
> 
> _[pkg/sql/logictest/testdata/logic_test/sequences, line 2076 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl3P8_EMqcAIIOr5Fa:-Msl3P8a2P6XWwaX7xmv:b-i0ha0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/db37312c44ebebe741629b7480a2f2bbaaf004a9/pkg/sql/logictest/testdata/logic_test/sequences#L2076)):_
> 
> > ```
> > SELECT last_value from customer_seq_check_cache_and_bounds_2
> > ----
> > -17
> > ```
> 
> `-17` is a pretty surprising result to get here. in postgres, this query returns `-1`

I think it is another issue that I am also confused about. When I use cockroach-v21.1.7.linux-amd64
```sql
drop sequence if exists test1;
create sequence test1 minvalue -1 maxvalue 2 start with 1 cache 5 increment by -2;
select nextval('test1');  -- 1
select nextval('test1');  -- -1
select nextval('test1');  -- nextval(): reached minimum value of sequence "test1" (-1)
select currval('test1');  -- -1
select last_value from test1; -- -17


drop sequence if exists test1;
create sequence test1 minvalue -1 maxvalue 2 start with 1 cache 5 increment by -2;
select nextval('test1');  -- 1
select nextval('test1');  -- -1
select currval('test1');  -- -1
select last_value from test1; -- -7
```

mosquito2333 commented:
The code has been modified, can you review it for me again? @ajwerner @rafiss

ajwerner reviewed (commented):
Nice, this is getting close. 

> Reviewed 1 of 4 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-:-MtJ8mrQ5c8-9TBto6V_:b-rn7qjg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mosquito2333, and @rafiss)

___
*[pkg/sql/alter_sequence.go, line 157 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl5sOXBNLdm0OWloXD:-MtJ8-u97uYlIZgN-iPC:b-pxlpb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/96acd7d758f351bc69dd307a52a4684bde10933d/pkg/sql/alter_sequence.go#L157)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should the "clear cache and record latest val" be done after writeSchemaChange?
</blockquote></details>

I also prefer this, but it's minor. 
___
*[pkg/sql/alter_sequence.go, line 122 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MtJ8QsQ3TzyQtWfYyYy:-MtJ8QsQ3TzyQtWfYyYz:b-ynzt9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/164538819943f434b9c4f412a9b331664ca8471b/pkg/sql/alter_sequence.go#L122)):*
> ```Go
> 	}
> 
> 	// If the sequenceVal exceeded the old MinValue or MaxValue, we must set sequenceVal to the last valid one.
> ```

nit: can you wrap your comments to 80?
___
*[pkg/sql/alter_sequence.go, line 128 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MtJ8IWi2J1mj568Ot5v:-MtJ8IWi2J1mj568Ot5w:b-2550y0) ([raw file](https://github.com/cockroachdb/cockroach/blob/164538819943f434b9c4f412a9b331664ca8471b/pkg/sql/alter_sequence.go#L128)):*
> ```Go
> 	hasRestart := false
> 	if hasRestart {
> 		sequenceVal = opts.Start - opts.Increment
> ```

I guess we should file an issue to support `RESTART` :sweat_smile: 



_Code quote:_
```Go
	// I think we should set the current value to the newStartValue if there is a start param in alter_sequence_stmt,
	// but in pg it needs RESTART. Now newStartValue has no effect on the current value before use RESTART.
	hasRestart := false
	if hasRestart {
		sequenceVal = opts.Start - opts.Increment
```
___
*[pkg/sql/sequence.go, line 158 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MtJ8c8t1Dui7cGlPcye:-MtJ8c8t1Dui7cGlPcyf:b-53suvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/164538819943f434b9c4f412a9b331664ca8471b/pkg/sql/sequence.go#L158)):*
> ```Go
> 		nextVal := preIncrement + opts.Increment
> 
> 		// The postIncrement value is in-bounds, this implies that nextVal and all interceding values are too.
> ```

nit: can you wrap the comments around here at 80?



_Code quote:_
```Go
// The postIncrement value is out-of-bounds, find the portion of the cached sequence which is valid, if any.
```
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1535 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-MtJ8ijQ0Q9pePnZfprq:-MtJ8ijQ0Q9pePnZfprr:b7vutr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/164538819943f434b9c4f412a9b331664ca8471b/pkg/sql/logictest/testdata/logic_test/sequences#L1535)):*
> ```
> #SELECT nextval('cache_test')
> #----
> #15
> ```

how come these tests are commented out?



_Code quote:_
```

#SELECT nextval('cache_test')
#----
#15
```


<!-- Sent from Reviewable.io -->

mosquito2333 commented on pkg/sql/alter_sequence.go:
```diff
@@ -112,34 +114,56 @@ func (n *alterSequenceNode) startExec(params runParams) error {
+			sequenceVal = oldStart - opts.Increment
+		} else {
+			// 2. The sequence is altered after it has ever been used.
+			currentVal, err := params.extendedEvalCtx.EvalContext.Sequence.GetLatestValueInSessionForSequenceByID(
```
I'm now confused about how to get the current value of the sequence

mosquito2333 replied:
Since the sequenceVal obtained from seqValueKey is the last value generated in the cache, and we need the current sequence value, I added a currentSeqKey to store it.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mosquito2333, and @rafiss)

___
*[-- commits, line 28 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/74321#-Msl3l1lDCfp4tB1xGcV):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the commit message should have a release note:
```
Release note (bug fix): A bug with ALTER SEQUENCE that allowed a sequence to return values
that are out-of-range has been fixed.
```
</blockquote></details>

Thanks rafi, indeed.

For bug fixes, it's also nice if the release note can include the version when the bug has been originally introduced, if known.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
>Although nextval is non transactional, we have some wonky behaviour where nextval is called on a sequence in the same transaction it is alterted or created.
When alter sequence, I need to modify sequenceVal, but it seems that alter_sequence_stmt opens a transaction that will compete with nextval() for sequenceVal, thus locking in the following two cases.
It is similar with https://github.com/cockroachdb/cockroach/issues/61425

This issue is fixed, do you plan to update this PR? I wouldn't mind take it over

mosquito2333 commented:
> 

I will update it.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=74321) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=74321) before we can accept your contribution.<br/><hr/>**liwen.sun@esgyn.cn** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=74321) it.</sub>

knz commented:
nit: before this merges, we'll want to review/expand the release note(s) and update the PR description at top.

mosquito2333 committed:
sql : alter sequence's increment and the results of 'select nextval()' were not as expected

If the sequenceVal exceeded the old MinValue or MaxValue, we must set sequenceVal to the last valid one.

   1. Values is just init, when the sequence is altered before it has ever been used.
   In this case, last valid sequenceVal is oldStart.
   Set the initial value to the oldStartValue(just like pg), when last valid sequenceVal in [newMinValue, newMaxValue].

   2. When the sequenceVal out of range result from increment, report an error.

Fixed #52552 #74127

Release note (bug fix): A bug with ALTER SEQUENCE that allowed a sequence to return values that are out-of-range has been fixed.

mosquito2333 commented:
> > Although nextval is non transactional, we have some wonky behaviour where nextval is called on a sequence in the same transaction it is alterted or created.
> > When alter sequence, I need to modify sequenceVal, but it seems that alter_sequence_stmt opens a transaction that will compete with nextval() for sequenceVal, thus locking in the following two cases.
> > It is similar with #61425
> 
> This issue is fixed, do you plan to update this PR? I wouldn't mind take it over

I'm sure you have a better solution, and my solution is not perfect.
Can you take it over?
